src/NVMLDevicePool.cpp

```diff
@@ -0,0 +1,367 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <cmath>
+
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <string>
+
+#include "Exception.hpp"
+#include "Agg.hpp"
+#include "Helper.hpp"
+#include "geopm_sched.h"
+
+#include "NVMLDevicePoolImp.hpp"
+
+namespace geopm
+{
+
+    const NVMLDevicePool &nvml_device_pool(const int num_cpu)
+    {
+        static NVMLDevicePoolImp instance(num_cpu);
+        return instance;
+    }
+
+    NVMLDevicePoolImp::NVMLDevicePoolImp(const int num_cpu)
+        : M_MAX_CONTEXTS(64)
+        , M_NUM_CPU(num_cpu)
+    {
+        nvmlReturn_t nvml_result;
+
+        //Initialize NVML
+        nvml_result = nvmlInit();
+        check_nvml_result(nvml_result, GEOPM_ERROR_RUNTIME, "NVMLDevicePool::" + std::string(__func__) +
+                          ": NVML failed to initialize.");
+
+        //Query number of NVML Accelerators
+        nvml_result = nvmlDeviceGetCount(&m_num_accelerator);
+        check_nvml_result(nvml_result, GEOPM_ERROR_RUNTIME, "NVMLDevicePool::" + std::string(__func__) +
+                          ": NVML failed to query device count.");
+
+        //Acquire device handle for each NVML Accelerator
+        m_nvml_device.resize(m_num_accelerator);
+        for (unsigned int accel_idx = 0; accel_idx < m_num_accelerator; ++accel_idx) {
+            nvml_result = nvmlDeviceGetHandleByIndex(accel_idx,
+                                                     &m_nvml_device.at(accel_idx));
+            check_nvml_result(nvml_result, GEOPM_ERROR_RUNTIME, "NVMLDevicePool::" + std::string(__func__) +
+                              ": NVML failed to get handle for accelerator " +
+                              std::to_string(accel_idx) + ".");
+        }
+    }
+
+    NVMLDevicePoolImp::~NVMLDevicePoolImp()
+    {
+        //Shutdown NVML
+        nvmlReturn_t nvml_result = nvmlShutdown();
+#ifdef GEOPM_DEBUG
+        if (nvml_result != NVML_SUCCESS) {
+            std::cerr << "NVMLDevicePool::" << __func__ <<  ": NVML failed to shutdown."
+                      << "  Error: " <<  nvmlErrorString(nvml_result) << std::endl;
+        }
+#endif
+    }
+
+    void NVMLDevicePoolImp::check_nvml_result(nvmlReturn_t nvml_result, int error, std::string message) const
+    {
+        if (nvml_result != NVML_SUCCESS) {
+            throw Exception(message + "  Error: " + nvmlErrorString(nvml_result), error, __FILE__, __LINE__);
```

Comment:
`__FILE__` and `__LINE__` should be parameters to this function (otherwise it will always print NVMLDevicePool.cpp:98)

Reply:
__LINE__ should definitely be a parameter, fixed in latest commit.  All NVML Handling is supposed to occur in this file, which leads to me wanting to keep __FILE__ as is.